tests/samples_test.py

```diff
@@ -0,0 +1,359 @@
+    for p, a in zip(samples, new_points):
+        assert p == pytest.approx(a)
+
+
+@pytest.mark.parametrize(
+    "nested_samples,num_per_level,start_level,end_level",
+    [
+        (
+            smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 0),
+            [1],
+            [0],
+            [1],
+        ),
+        (
+            smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 1),
+            [1, 2],
+            [0, 1],
+            [1, 3],
+        ),
+        (
+            smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 2),
+            [1, 2, 2],
+            [0, 1, 3],
+            [1, 3, 5],
+        ),
+        (
+            smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 3),
+            [1, 2, 2, 4],
+            [0, 1, 3, 5],
+            [1, 3, 5, 9],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 0),
+            [1],
+            [0],
+            [1],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 1),
+            [1, 2],
+            [0, 1],
+            [1, 3],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 2),
+            [1, 2, 2],
+            [0, 1, 3],
+            [1, 3, 5],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 3),
+            [1, 2, 2, 4],
+            [0, 1, 3, 5],
+            [1, 3, 5, 9],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 4),
+            [1, 2, 2, 4, 0],
+            [0, 1, 3, 5, 9],
+            [1, 3, 5, 9, 9],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            [1, 2, 2, 4, 0, 8],
+            [0, 1, 3, 5, 9, 9],
+            [1, 3, 5, 9, 9, 17],
+        ),
+        (
+            smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 0),
+            [1],
+            [0],
+            [1],
+        ),
+        (
+            smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 1),
+            [1, 2],
+            [0, 1],
+            [1, 3],
+        ),
+        (
+            smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 2),
+            [1, 2, 6],
+            [0, 1, 3],
+            [1, 3, 9],
+        ),
+    ],
+    ids=nested_sample_ids,
+)
+def test_nested_levels(nested_samples, num_per_level, start_level, end_level):
+    """test number of points per level, start level indexes, and end level indexes"""
+    assert numpy.array_equal(nested_samples.num_per_level, num_per_level)
+    assert numpy.array_equal(nested_samples.start_level, start_level)
+    assert numpy.array_equal(nested_samples.end_level, end_level)
+    for i in range(len(start_level)):
+        points = nested_samples[start_level[i] : end_level[i]]
+        assert numpy.allclose(nested_samples.level(i), points)
```

Comment:
You can use enumerate and zip here:
```suggestion
    for level, (start, end) in enumerate(zip(nested_samples.start_level, nested_samples.end_level)):
        assert numpy.allclose(nested_samples.level(level), nested_samples[start:end])
```